File: selection/py/quick_select.py
# Finds the kth largest element in the array by partially sorting the array
# At the end, all k smallest elements are placed at array[:k + 1]
# All elements greater than the kth largest element are placed at array[k + 1:]
# Time:  O(n)

def insertion_sort(array, left, right):
    for i in range(left + 1, right + 1):
        x = array[i]
        j = i - 1
        while j >= left and array[j] > x:
            array[j + 1] = array[j]
            j -= 1
        array[j + 1] = x

def median3(array, left, right):
    center = (left + right) / 2
    temp = [array[left], array[center], array[right]]
    insertion_sort(temp, 0, 2)
    array[left], array[right], array[center] = temp
    return array[right]

def partition(array, left, right):
    pivot = median3(array, left, right)
    i, j = left + 1, right - 1
    while True:
        while array[i] < pivot:
            i += 1
        while pivot < array[j]:
            j -= 1
        if i >= j:
            array[i], array[right] = array[right], array[i]
            return i
        array[i], array[j] = array[j], array[i]
        i, j = i + 1, j - 1

def q_select(array, left, right, rank):
    if right - left < 10:
        insertion_sort(array, left, right)
        return
    center = partition(array, left, right)
    if rank < center:
        q_select(array, left, center - 1, rank)
    elif center < rank:
        q_select(array, center + 1, right, rank)

def quick_select(array, rank):
    q_select(array, 0, len(array) - 1, rank - 1)
    return array[rank - 1]

# Binary search on an array can usually be implemented non-recursively
def q_select_loop(array, left, right, rank):
    while right - left > 10:
        center = partition(array, left, right)
        if center == rank:
            return
        elif center < rank:
            left = center + 1
        else:
            right = center - 1
    insertion_sort(array, left, right)

File: selection/java/QuickSelect.java
import java.util.Random;

public class QuickSelect implements SelectionAlgorithm
{
	private Random rand = new Random();
	private void insertion_sort(int[] array, int left, int right)
	{
		int j, x;
		for(int i = left + 1; i < right + 1; i++)
		{
			x = array[i];
			j = i - 1;
			for(j = i - 1; j >= left && array[j] > x; j--)
				array[j + 1] = array[j];
			array[j + 1] = x;
		}
	}
	private int median3(int[] array, int left, int right)
	{
		int center = (left + right) / 2;
		int tmp[] = { array[left], array[center], array[right] };
		insertion_sort(tmp, 0, 2);
		array[left] = tmp[0];
		array[center] = tmp[2];
		array[right] = tmp[1];
		return array[right];
	}
	private int partition(int[] array, int left, int right)
	{
		int pivot = median3(array, left, right);
		int i = left, j = right - 1;
		while(true)
		{
			while(array[i] < pivot) i++;
			while(pivot < array[j]) j--;
			if(i >= j)
			{
				array[i] = array[right] ^ array[i] ^ (array[right] = array[i]);
				return i;
			}
			array[i] = array[j] ^ array[i] ^ (array[j] = array[i]);
			i++; j--;
		}
	}
	private void q_select(int[] array, int left, int right, int rank)
	{
		if(right - left < 10)
		{
			insertion_sort(array, left, right);
			return;
		}
		int center = partition(array, left, right);
		if(rank < center)
			q_select(array, left, center - 1, rank);
		else if(rank > center)
			q_select(array, center + 1, right, rank);
	}
	public int select(int[] array, int rank)
	{
		q_select(array, 0, array.length - 1, rank - 1);
		return array[rank - 1];
	}
}
